$OpenBSD: patch-src_libpiano_response_c,v 1.3 2013/09/02 11:58:40 dcoppa Exp $

piano: replace station with same id.
(upstream git commit 6e82d7e6852fdac4a178445b3e2bcdb6d4e0fa63)

--- src/libpiano/response.c.orig	Sun May 19 12:58:18 2013
+++ src/libpiano/response.c	Mon Sep  2 11:28:04 2013
@@ -438,15 +438,31 @@ PianoReturn_t PianoResponse (PianoHandle_t *ph, PianoR
 
 			PianoJsonParseStation (result, tmpStation);
 
-			/* start new linked list or append */
 			if (ph->stations == NULL) {
 				ph->stations = tmpStation;
 			} else {
-				PianoStation_t *curStation = ph->stations;
+				PianoStation_t *curStation = ph->stations, *prevStation = NULL;
 				while (curStation->next != NULL) {
+					/* replace if station with same id exists already */
+					if (strcmp (curStation->id, tmpStation->id) == 0) {
+						if (prevStation == NULL) {
+							ph->stations = tmpStation;
+						} else {
+							prevStation->next = tmpStation;
+						}
+						tmpStation->next = curStation->next;
+
+						PianoDestroyStation (curStation);
+						free (curStation);
+						break;
+					}
+					prevStation = curStation;
 					curStation = curStation->next;
 				}
-				curStation->next = tmpStation;
+				/* append otherwise */
+				if (tmpStation->next == NULL) {
+					curStation->next = tmpStation;
+				}
 			}
 			break;
 		}
